java/c/src/main/java/org/apache/arrow/c/BufferImportTypeVisitor.java

```diff
@@ -116,7 +119,7 @@ private ArrowBuf importFixedBytes(ArrowType type, int index, long bytesPerSlot)
   }
 
   private ArrowBuf importOffsets(ArrowType type, long bytesPerSlot) {
-    final long capacity = bytesPerSlot * (fieldNode.getLength() + 1);
+    final long capacity = bytesPerSlot * (fieldNode.getLength() + arrowArrayOffset + 1);
```

Comment:
Why are we adjusting capacity here?

Reply:
Right. This is not needed.

Reply:
Oh wait, this is needed because this fails the `integration_tests.py` 

```bash
======================================================================
ERROR: test_binary_slice_array (__main__.TestPythonIntegration)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "Data.java", line 319, in org.apache.arrow.c.Data.importVector
java.lang.java.lang.IndexOutOfBoundsException: java.lang.IndexOutOfBoundsException: index (24) must not be greater than size (16)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "Data.java", line 319, in org.apache.arrow.c.Data.importVector
Exception: Java Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "integration_tests.py", line 202, in test_binary_slice_array
    self.round_trip_array(lambda: sliced_array)
  File "integration_tests.py", line 160, in round_trip_array
    self.bridge.python_to_java_array(original_arr, dictionary_provider) as vector:
  File "integration_tests.py", line 111, in python_to_java_array
    return self.java_c.Data.importVector(self.java_allocator, c_array, c_schema, dictionary_provider)
java.lang.java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: Could not load buffers for field : Binary. error message: index (24) must not be greater than size (16)

======================================================================
ERROR: test_string_slice_array (__main__.TestPythonIntegration)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "Data.java", line 319, in org.apache.arrow.c.Data.importVector
java.lang.java.lang.IndexOutOfBoundsException: java.lang.IndexOutOfBoundsException: index (24) must not be greater than size (16)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "Data.java", line 319, in org.apache.arrow.c.Data.importVector
Exception: Java Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "integration_tests.py", line 196, in test_string_slice_array
    self.round_trip_array(lambda: sliced_array)
  File "integration_tests.py", line 160, in round_trip_array
    self.bridge.python_to_java_array(original_arr, dictionary_provider) as vector:
  File "integration_tests.py", line 111, in python_to_java_array
    return self.java_c.Data.importVector(self.java_allocator, c_array, c_schema, dictionary_provider)
java.lang.java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: Could not load buffers for field : Utf8. error message: index (24) must not be greater than size (16)
```

Reply:
If we don't add the `arrowArrayOffset` then we get 4 slots, which are the first four slots, but what we need is not the first 4 slots, but the from `arrowArrayOffset` + 4 slots. And then we have to get the adjusted offset from this since Java doesn't need to keep the whole offset buffer like C++.  Am I missing something here?

Reply:
But we should be adjusting the start of the buffer, not the capacity?

Reply:
Or does the length have the offset pre-applied?

Reply:
`fieldNode.getLength()=3` as expected from slicing. Which means, we exactly don't know which 3 values to read here. I thought, since there is slicing, we need to read more from that buffer meaning, additional `arrowArrayOffset` amount. 

> But we should be adjusting the start of the buffer, not the capacity?

I completely agree, but I couldn't figure out a way to do that from the current code, we cannot provide a starting point, but a capacity to read from the beginning. Should we improve the memory read component?

Reply:
?? Just read the full buffer then slice.

Reply:
Right, in the C data interface `ArrowArray.Snapshot` only provides us `length` and the `offset`, other than that we have no information to determine the size of the full buffer. So what we read now has the capacity of `length + offset`. Am I missing something here?

Reply:
Right, but here I see only the full buffer, no slice

Reply:
Oh that's right. I will move the redundant logic in utf8 and binary type to here.